### PR TITLE
Improve FS by dropping the encryption_secret after use

### LINKED ISSUE #91: Enforce FS for secret tree
Make sure `encryption_secret` is deleted as soon as the secret tree is created.

### PR DESCRIPTION
**Pull request by kkohbrok**
Fixes #91 by introducing a dedicated function to retrieve the `encryption_secret` from an `EpochSecrets` object. The function removes the `Secret` such that ownership can be passed to the `SecretTree` constructor, at the end of which the Secret is dropped.

### CODE REVIEW

File: src/schedule/mod.rs [hunk 1]
```diff
@@ -76,7 +85,7 @@ impl Default for EpochSecrets {
     fn default() -> Self {
         let welcome_secret = Secret::new_empty_secret();
         let sender_data_secret = Secret::new_empty_secret();
-        let encryption_secret = Secret::new_empty_secret();
+        let encryption_secret = RefCell::new(Some(Secret::new_empty_secret()));
```
Comment (raphaelrobert) [inline src/schedule/mod.rs:88]: Is there a reason you need the `RefCell` here? Why not just an `Option` that you can later consume with`.take()`?

File: src/schedule/mod.rs [hunk 2]
```diff
@@ -139,13 +149,26 @@ impl EpochSecrets {
...
+            Some(es) => es,
+            None => return Err(KeyScheduleError::SecretReuseError),
+        };
+        // Pass ownership to the `SecretTree` constructor, so that the
+        // `encryption_secret` is dropped afterwards.
+        Ok(encryption_secret)
+    }
 }
 
 impl Codec for EpochSecrets {
```
Comment (raphaelrobert) [inline src/schedule/mod.rs:167]: Do we need the encoding at all btw? I think all we need is persistence going forward.

File: src/schedule/mod.rs (hunk 1 above)
Comment (kkohbrok) [reply src/schedule/mod.rs:88]: Good point. That means we need to make `EpochSecrets` `mut`, but that should be just fine.

File: src/schedule/mod.rs (hunk 2 above)
Comment (kkohbrok) [reply src/schedule/mod.rs:167]: Doesn't seem like it. I had already closed #207, but I just reopened it, so the issue is visible. I won't tackle it in this PR, though.

File: src/schedule/mod.rs [hunk 3]
```diff
@@ -152,9 +149,9 @@ impl EpochSecrets {
 
     /// Remove the `encryption_secret` from the `EpochSecrets`, replacing it
     /// with `None` and return it.
-    pub fn remove_encryption_secret(&self) -> Result<Secret, KeyScheduleError> {
+    pub fn remove_encryption_secret(&mut self) -> Result<Secret, KeyScheduleError> {
```
Comment (raphaelrobert) [inline src/schedule/mod.rs:152]: Maybe `consume_encryption_secret` is more appropriate, given that the secret is returned?

**Comment by kkohbrok** [comment]
I moved the actual consumption of the `encryption_secret` into a function of `EpochSecrets` to guarantee that it's actually dropped and to reduce the amount of exposure that secrets get.

**Review by raphaelrobert — APPROVED** [review]

**Pull request merged**